src/main/java/com/ebivariation/contigalias/service/AssemblyService.java

```diff
@@ -16,36 +16,123 @@
+        if (fetchAssembly.isPresent()) {
+            AssemblyEntity asm = fetchAssembly.get();
             List<ChromosomeEntity> chromosomes = asm.getChromosomes();
             if (chromosomes != null) {
                 chromosomes.forEach(chr -> chr.setAssembly(null));
             }
+            // Setting this to null to remove inconsistency in results
+            // When fetching from db bring no results, chromosomes is an empty list
+            // whereas here chromosomes is value
+            else {
+                asm.setChromosomes(new LinkedList<>());
+            }
+            return fetchAssembly;
+        }
+        return Optional.empty();
+    }
+
+    public Optional<AssemblyEntity> fetchAndInsertAssembly(String accession) throws IOException {
+
+        if (repository.findAssemblyEntityByAccession(accession).isPresent()) {
+            throw duplicateAssemblyInsertionException;
+        }
+
+        Optional<AssemblyEntity> fetchAssembly = dataSource.getAssemblyByAccession(accession);
+        fetchAssembly.ifPresent(this::insertAssembly);
+        return fetchAssembly;
+    }
+
+    public Optional<AssemblyEntity> getAssemblyByAccession(String accession) {
+        Optional<AssemblyEntity> assembly = repository.findAssemblyEntityByAccession(accession);
+        assembly.ifPresent(asm -> {
+            List<ChromosomeEntity> chromosomes = asm.getChromosomes();
+            if (chromosomes != null && chromosomes.size() > 0) {
+                chromosomes.forEach(chr -> chr.setAssembly(null));
+            }
         });
         return assembly;
     }
 
+    public void insertAssembly(AssemblyEntity entity) {
+        // Limit cache size to 10 assemblies
+        while (repository.countByIdNotNull() >= 10) {
+            repository.findTopByIdNotNullOrderById().ifPresent(it -> repository.deleteById(it.getId()));
+        }
+
+        if (isEntityPresent(entity)) {
+            throw duplicateAssemblyInsertionException;
+        } else {
+            repository.save(entity);
+        }
+    }
+
+    public void deleteAssembly(AssemblyEntity entity) {
+        if (isEntityPresent(entity)) {
+            repository.delete(entity);
+        }
+    }
+
+    public boolean isEntityPresent(AssemblyEntity entity) {
+        Optional<AssemblyEntity> existingAssembly = repository.findAssemblyEntityByGenbankOrRefseq(
+                // Setting to invalid prevents finding random accessions with null GCA/GCF
+                entity.getGenbank() == null ? "##########" : entity.getGenbank(),
+                entity.getRefseq() == null ? "##########" : entity.getRefseq());
+        return existingAssembly.isPresent();
+    }
+
+    public void fetchAndInsertAssembly(List<String> accessions) {
+        accessions.forEach(it -> executor.submit(() -> {
+            try {
+                this.fetchAndInsertAssembly(it);
+            } catch (IOException e) {
+                e.printStackTrace();
```

Comment:
either rethrow, or log the exception and rethrow.

EDIT: oh wait this is a thread executor. I understand that you may not be able to rethrow. The point about logging still applies. I have to study carefully what to do in this case.